### PR TITLE
Vault setup with Spire OIDC

### PR DESCRIPTION
**Pull request by sudo-bmitch**
This one changes the key storage from a secret to inside of vault. Vault is setup with the transit plugin to allow authorized users to sign data without ever giving out the private key. And for authentication, this adds Spire with OIDC. Spire authenticates the node is in the cluster, and the request is coming from the Tekton chains service account. That allows the chains operator to get a short lived SVID from Spire which it uses to authenticate to Vault using OIDC.

Spire is considering taking over the helm chart. See ([Spire #2652](https://github.com/spiffe/spire/issues/2652))

Signed-off-by: Brandon Mitchell <git@bmitch.net>

### CODE REVIEW

**Comment by joshualucas84** [comment]
@sudo-bmitch Brandon so when i run make setup-vault i get the following error.

image

**Comment by sudo-bmitch** [comment]
> @sudo-bmitch Brandon so when i run make setup-vault i get the following error.
> 
> image

Hopefully that's just a race condition. Check if the pods and services came up with `kubectl get all -n spire`.

**Comment by joshualucas84** [comment]
> > @sudo-bmitch Brandon so when i run make setup-vault i get the following error.
> > image
> 
> Hopefully that's just a race condition. Check if the pods and services came up with `kubectl get all -n spire`.

It was I hadn't ran bash platform/20-spire-setup.sh yet, wonder if we should add that to vault setup as a prereq?

**Comment by sudo-bmitch** [comment]
> It was I hadn't ran bash platform/20-spire-setup.sh yet, wonder if we should add that to vault setup as a prereq?

@joshualucas84 well there's a face palm moment. I'll get that added to the Makefile. Just need work some git surgery to shuffle around some other commits to get them out of this branch first.

**Comment by sudo-bmitch** [comment]
> > It was I hadn't ran bash platform/20-spire-setup.sh yet, wonder if we should add that to vault setup as a prereq?
> 
> @joshualucas84 well there's a face palm moment. I'll get that added to the Makefile. Just need work some git surgery to shuffle around some other commits to get them out of this branch first.

Added another entry to the Makefile and rebased on main. There's still some work left to make this happen. I can get some debug code over to you to try out spire + vault + cosign. What I'm looking at right now is whether Cosign or Chains should be updated to allow the Chains KMS integration to use the Spire access to generate a vault token on the fly, or if that's too specific and we need to add some kind of extensibility interface.

**Comment by sudo-bmitch** [comment]
This one is still in work upstream. Sigstore changes are done, and chains work is developed, just finishing the testing code on that so we can get it merged.

File: platform/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+certs
+components/vault/root-token
+components/vault/unseal-key
```
Comment (bradbeck) [inline platform/.gitignore:3]:
```suggestion
components/vault/unseal-key

```

File: platform/components/tekton/chains/patch_spire.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+              "path": "/run/spire/sockets",
+              "type": "DirectoryOrCreate"
+            },
+            "name": "spire-agent-socket"
+          }
+        ]
+      }
+    }
+  }
+}
```
Comment (bradbeck) [inline platform/components/tekton/chains/patch_spire.json:55]:
```suggestion
}

```

File: platform/vault/values.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+injector:
+  enabled: false
+
+server:
+  affinity: ""
+
```
Comment (bradbeck) [inline platform/vault/values.yaml:6]:
```suggestion

```

File: platform/components/vault/values.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  #   enabled: true
+  #   config: |
+  #     listener "tcp" {
+  #       address = "[::]:8200"
+  #       cluster_address = "[::]:8201"
+  #     }
+
+  #     storage "file" {
+  #       path = "/vault/data"
+  #     }
```
Comment (bradbeck) [inline platform/components/vault/values.yaml:31]:
```suggestion
  #     }

```

File: platform/components/vault/values.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  # standalone:
+  #   enabled: true
+  #   config: |
+  #     listener "tcp" {
+  #       address = "[::]:8200"
+  #       cluster_address = "[::]:8201"
+  #     }
+  #     storage "file" {
+  #       path = "/vault/data"
+  #     }
```
[the remark this answers is not in the document]
Comment (sudo-bmitch) [reply platform/components/vault/values.yaml:29]: I was looking at that warning, but not sure of the best way to clear it. It should be indented under server, same level as affinity. I'll give the yaml linter a check to see if there's a better way I should be doing it.

File: platform/06-setup-certs.sh [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+# create the CA key/cert
+[ -d "${GIT_ROOT}/platform/certs/ca" ] || mkdir -p "${GIT_ROOT}/platform/certs/ca"
+ca_key="${GIT_ROOT}/platform/certs/ca/ca-key.pem"
+ca_cert="${GIT_ROOT}/platform/certs/ca/ca.pem"
+[ -f "$ca_key" ] || openssl genrsa -out "$ca_key" 4096
+openssl req -new -x509 \
+  -key "$ca_key" \
+  -days 3650 \
+  -subj "/CN='Demo CA'" \
+  -addext "basicConstraints=CA:true" \
```
[the remark this answers is not in the document]
Comment (bradbeck) [reply platform/06-setup-certs.sh:19]: Looks like we need at least `LibreSSL 3.1.0`.

File: platform/20-spire-setup.sh [hunk 7]
```diff
@@ -1,88 +1,46 @@
...
-  -f "${QUICKSTART_DIR}/spire-bundle-configmap.yaml" \
-  -f "${QUICKSTART_DIR}/server-cluster-role.yaml"
+kubectl create namespace spire --dry-run=client -o yaml | kubectl apply -f -
 
-# Create Server Configmap.
-kubectl apply \
-  -f "${QUICKSTART_DIR}/server-configmap.yaml" \
-  -f "${QUICKSTART_DIR}/server-statefulset.yaml" \
-  -f "${QUICKSTART_DIR}/server-service.yaml"
+helm repo add sudo-bmitch https://sudo-bmitch.github.io/helm-charts
```
Comment (bradbeck) [inline platform/20-spire-setup.sh:26]: This line fails for me:
```bash
+ helm repo add sudo-bmitch https://sudo-bmitch.github.io/helm-charts
Error: looks like "https://sudo-bmitch.github.io/helm-charts" is not a valid chart repository or cannot be reached: Get "https://sudo-bmitch.github.io/helm-charts/index.yaml": dial tcp [2606:50c0:8002::153]:443: connect: no route to host
```

**Review by bradbeck — DISMISSED** [review]

File: platform/06-setup-certs.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (sudo-bmitch) [reply platform/06-setup-certs.sh:19]: This is one of the areas that will be helped with cert manager.

**Comment by rgreinho** [comment]
Running `make quickstart` produces the following error for me:

```bash
Configuring helm...
"stable" already exists with the same configuration, skipping
"bitnami" already exists with the same configuration, skipping
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "bitnami" chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈Happy Helming!⎈
bash platform/06-setup-certs.sh
unknown option -addext
```

It looks like the `openssl req` subcommand on OSX (`macOS 12.3.1`) does not have the `-addext` flag. 

My openssl version is:

```bash
$ openssl version
LibreSSL 2.8.3
```

**Comment by sudo-bmitch** [comment]
> ```
> bash platform/06-setup-certs.sh
> unknown option -addext
> ```

I think Brad ran into that too, and has a resolution in his comment: https://github.com/buildsec/frsca/pull/94#discussion_r887209446

This is a short term gotcha, I'm planning to replace that script with cert-manager.

**Review by rgreinho — DISMISSED** [review]
I got it to deploy successfully and I was able to run the buildpacks pipeline without any issue 👌

File: platform/20-spire-setup.sh [hunk 8]
```diff
@@ -1,88 +1,46 @@
...
+kubectl rollout status -n spire daemonset/spire-agent
 
 # Register Workloads.
 spire_apply \
-  -spiffeID spiffe://example.org/ns/spire/sa/spire-agent \
-  -selector k8s_sat:cluster:demo-cluster \
-  -selector k8s_sat:agent_ns:spire \
-  -selector k8s_sat:agent_sa:spire-agent \
+  -spiffeID spiffe://example.org/ns/spire/node/ssf \
+  -selector k8s_psat:cluster:ssf \
```
Comment (bradbeck) [inline platform/20-spire-setup.sh:38]: Seems like we should change `ssf` to `frsca` or something else.

File: platform/20-spire-setup.sh [hunk 9]
```diff
@@ -1,88 +1,46 @@
...
+kubectl rollout status -n spire statefulset/spire-server
+kubectl rollout status -n spire daemonset/spire-agent
 
 # Register Workloads.
 spire_apply \
-  -spiffeID spiffe://example.org/ns/spire/sa/spire-agent \
-  -selector k8s_sat:cluster:demo-cluster \
-  -selector k8s_sat:agent_ns:spire \
-  -selector k8s_sat:agent_sa:spire-agent \
+  -spiffeID spiffe://example.org/ns/spire/node/ssf \
```
Comment (bradbeck) [inline platform/20-spire-setup.sh:37]: Seems like we should change `ssf` to `frsca` or something else.

File: platform/26-vault-setup.sh [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+kubectl -n vault create configmap ssf-certs --from-file=ssf.pem="${GIT_ROOT}/platform/certs/ssf.pem" --dry-run=client -o=yaml | kubectl apply -f -
+
+COSIGN_KEY=$(kubectl get secret signing-secrets -n tekton-chains -o jsonpath='{.data.cosign\.key}' 2> /dev/null || true)
+if [ -n "${COSIGN_KEY}" ]; then
+  # remove signing-secrets if created by cosign
+  kubectl -n tekton-chains delete secret signing-secrets
+fi
+( kubectl -n tekton-chains create secret generic signing-secrets \
+    --from-file=cosign.pub="${GIT_ROOT}/platform/certs/ssf.pem" \
```
Comment (bradbeck) [inline platform/26-vault-setup.sh:94]: Seems like we should change `ssf` to `frsca` or something else.

File: platform/components/spire/values.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
+server:
+  trustDomain: "example.org"
+  clusterName: ssf
```
Comment (bradbeck) [inline platform/components/spire/values.yaml:3]: Seems like we should change `ssf` to `frsca` or something else.

File: platform/components/tekton/chains/patch_config_kms.yaml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+data:
+  artifacts.oci.signer: kms
+  artifacts.taskrun.signer: kms
+  signers.kms.kmsref: "hashivault://ssf"
```
Comment (bradbeck) [inline platform/components/tekton/chains/patch_config_kms.yaml:4]: Seems like we should change `ssf` to `frsca` or something else.

**Comment by sudo-bmitch** [comment]
@bradbeck ssf references have been updated to frsca.

**Review by bradbeck — APPROVED** [review]

**Review by rgreinho — APPROVED** [review]

**Pull request merged**